### PR TITLE
Deprecate PSR-0 class loading

### PR DESCRIPTION
**Pull request by laoneo**
As PSR-0 is deprecated, this PR offers to register namespaces also for PSR-4. The plan is to remove support for PSR-0 in J4 probably.

Maintainer review.

Ping @wilsonge

### CODE REVIEW

**Comment by mbabker** [comment]
Why can't both be supported?  True that PSR-0 is deprecated but that doesn't mean it is unusable.  When I see a platform like Composer deprecate/drop PSR-0 support I might find this acceptable but for now I don't see any reason to do anything but add a deprecation notice and leave it be until the greater PHP ecosystem drops it in full.

Also, this change will 100% break the patch tester component because it does use the PSR-0 autoloader as I've namespaced the component classes.  The only way to therefore create a 3.x/4.x compatible build is to ship my own Composer autoloader since there won't be a path that allows me to load namespaced code in JLoader compatible with both versions.

**Comment by wilsonge** [comment]
You cannot use both PSR-4 and PSR-0 code at the same time. So are we going to be stuck using PSR-0 code for another 5 years until composer removes support?

**Comment by mbabker** [comment]
You can use both.  Look at our Framework packages as an example.  Are you saying I cannot install `joomla/application` and `joomla/session` together because the former uses PSR-4 and the latter PSR-0?

**Comment by mbabker** [comment]
What you cannot do is register a path that's designed for PSR-4 support to a PSR-0 autoloader or vice versa.  But you can most assuredly have an autoloader for each variant.

**Comment by wilsonge** [comment]
No but you must register each namespace with the relevant autoloader. Are we really going to introduce a `registerPSR4Namespace` method to register those namespaces differently.

**Comment by mbabker** [comment]
I would suggest for 4.0 that you support both.  Otherwise you're making the 4.0 upgrade path more complicated for the sake of it.

**Comment by mbabker** [comment]
Either way I will now update the patch tester to isolate itself from the core autoloader to prevent possible unneeded breakage either now or in the future.

**Comment by wilsonge** [comment]
img_5991

Fiiiiine

**Comment by laoneo** [comment]
When this one will be accepted PR #13673 needs to be adjusted as well.

File: libraries/loader.php [hunk 1]
```diff
@@ -380,18 +382,21 @@ public static function registerAlias($alias, $original, $version = false)
 	/**
 	 * Register a namespace to the autoloader. When loaded, namespace paths are searched in a "last in, first out" order.
 	 *
+	 * This function will be changed in J4 to support PSR-4 namespace registering.
```
Comment (wilsonge) [inline libraries/loader.php]: Can you mark this was `@note` please

File: libraries/loader.php [hunk 2]
```diff
@@ -152,11 +152,13 @@ public static function getDeprecatedAliases()
...
 	 *
+	 * @param   string   $type  Defines the type of namespace, can be prs0 or psr4.
+	 *
 	 * @return  array  The array of namespace => path values for the autoloader.
 	 *
 	 * @since   12.3
 	 */
-	public static function getNamespaces()
+	public static function getNamespaces($type = 'psr0')
 	{
```
Comment (wilsonge) [inline libraries/loader.php]: Can we add an `InvalidArgumentException` if the type is not psr0 or psr4 please?

File: libraries/loader.php [hunk 3]
```diff
@@ -380,18 +382,21 @@ public static function registerAlias($alias, $original, $version = false)
...
 	 *
 	 * @return  void
 	 *
 	 * @throws  RuntimeException
 	 *
 	 * @since   12.3
 	 */
-	public static function registerNamespace($namespace, $path, $reset = false, $prepend = false)
+	public static function registerNamespace($namespace, $path, $reset = false, $prepend = false, $type = 'psr0')
 	{
```
Comment (wilsonge) [inline libraries/loader.php]: Can we add an `InvalidArgumentException` if the type is not psr0 or psr4 please?

**Review by wilsonge — CHANGES_REQUESTED** [review]

File: libraries/loader.php [hunk 4]
```diff
@@ -457,18 +462,79 @@ public static function setup($enablePsr = true, $enablePrefixes = true, $enableC
...
+			$className = $class;
+		}
+
+		$classPath .= $className . '.php';
+
+		// Loop through registered namespaces until we find a match.
+		foreach (self::$namespaces['psr4'] as $ns => $paths)
+		{
+			$nsPath = trim(str_replace('\\', DIRECTORY_SEPARATOR, $ns), DIRECTORY_SEPARATOR);
+			if (strpos($class, $ns) === 0)
```
Comment (zero-24) [inline libraries/loader.php]: Please add a clean line before the if

File: libraries/loader.php [hunk 5]
```diff
@@ -152,14 +152,18 @@ public static function getDeprecatedAliases()
...
 	 * @return  array  The array of namespace => path values for the autoloader.
 	 *
 	 * @since   12.3
 	 */
 	public static function getNamespaces($type = 'psr0')
 	{
+		if ($type != 'psr0' && $type != 'psr4')
+		{
+			throw new InvalidArgumentException('Type needs to be prs0 or psr4!');
+		}
```
Comment (zero-24) [inline libraries/loader.php]: Here please too :)

**Review by zero-24 — COMMENTED** [review]
Here should be a clean line after the closing } and before the return

**Comment by zero-24** [comment]
Travis error:

FILE: /home/travis/build/joomla/joomla-cms/libraries/loader.php
--------------------------------------------------------------------------------
FOUND 1 ERROR(S) AFFECTING 1 LINE(S)
--------------------------------------------------------------------------------
 391 | ERROR | Parameters must appear immediately after the comment
--------------------------------------------------------------------------------
UPGRADE TO PHP_CODESNIFFER 2.0 TO FIX ERRORS AUTOMATICALLY
--------------------------------------------------------------------------------

**Review by wilsonge — APPROVED** [review]

**Comment by wilsonge** [comment]
Merging on review, given this is never going to get tests and we need it to progress in the sprint

**Comment by zero-24** [comment]
I have just fixed some smal CS Issues introduced from here ;)
https://github.com/joomla/joomla-cms/commit/f0d927051091a6725cc6bfe1e3b5ba13b81b8b4c

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**